Add BackendTLSPolicy support to Gateway API

youngnick opened the pull request:
This commit adds BackendTLSPolicy reconciliation (only) to the Gateway API reconciler.

This commit adds:
* watch BackendTLSPolicy resources
* Calculate their correct status and update it.

In order to get the BackendTLSPolicy status calculations correct, it turns out we need to pre-process all the BackendTLSPolicy objects, and build maps keyed via their targetRefs (the Services that they target). This is because of the upstream conflict resolution requirements for Policy objects, where conflicts need to be _resolved_ based on what Service (and port name) the Policy targets, and _reported_ based on the Gateway that they are used in.

The calculation for this is added in
`operator/pkg/gateway-api/helpers/backendtlspolicies.go`, and has both tests and documentation.

The actual status updating is performed in
`operator/pkg/gateway-api/gateway_reconcile.go`, in the `setBackendTLSPolicyStatuses` method.

This commit also changes the Gateway reconciliation test framework slightly so that all Gateways are reconciled for each test before Routes and other objects are checked.

This allows multiparent status checks to work.

This also removes the filtering of Routes, so all Routes in the input for the reconciler test must be present in the output to be checked. This change forces us to ensure that things that should have _no_ status updates don't get changed.

As part of making this change, I also found some tests (httproute-hostname-intersection, httproute-listener-hostname-matching, and httproute-listener-port-matching) that were not checking the correct Gateway status, that we all missed. I fixed those as well.

```release-note
Gateway API support now includes BackendTLSPolicy features
```

youngnick committed:
gateway-api: Add BackendTLSPolicy reconciliation to Gateway API

This commit adds BackendTLSPolicy reconciliation (only)
to the Gateway API reconciler.

This commit adds:
* watch BackendTLSPolicy objects
* watch ConfigMap objects used in BackendTLSPolicy objects
* Calculate their correct status and update it.

In order to get the BackendTLSPolicy status calculations
correct, it turns out we need to pre-process all the
BackendTLSPolicy objects, and build maps keyed via their
targetRefs (the Services that they target). This is because
of the upstream conflict resolution requirements for Policy
objects, where conflicts need to be _resolved_ based on
what Service (and port name) the Policy targets, and
_reported_ based on the Gateway that they are used in.

The calculation for this is added in
`operator/pkg/gateway-api/helpers/backendtlspolicies.go`,
and has both tests and documentation.

The actual status updating is performed in
`operator/pkg/gateway-api/gateway_reconcile.go`, in the
`setBackendTLSPolicyStatuses` method.

This commit also changes the Gateway reconciliation test framework
slightly so that all Gateways are reconciled for each test
before Routes and other objects are checked.

This allows multiparent status checks to work.

This also removes the filtering of Routes, so all Routes in
the input for the reconciler test must be present in the output
to be checked. This change forces us to ensure that things that
should have _no_ status updates don't get changed.

As part of making this change, I also found some tests
(httproute-hostname-intersection, httproute-listener-hostname-matching,
and httproute-listener-port-matching) that were not checking the
correct Gateway status, that we all missed. I fixed those as
well.

Further commits will add support for the ingestion
and translation steps.

Signed-off-by: Nick Young <nick@isovalent.com>

sathieu commented:
xref https://github.com/cilium/cilium/issues/20960 (should it be fixed by this PR?)

youngnick commented:
Part of it will be fixed, yes. But This one has proven _way_ more difficult than I thought, still working on it.

youngnick committed:
gateway-api: Add ingestion for BackendTLSPolicy

This commit adds ingestion support for BackendTLSPolicy objects.

The Gateway API model now stores TLS config for relevant backends
when there is a valid BackendTLSPolicy present.

Signed-off-by: Nick Young <nick@isovalent.com>

youngnick committed:
secretsync: Add synchronization of relevant ConfigMaps

This commit gives the secret-sync Cell the ability to
synchronize ConfigMaps into Secrets in the configured
secrets namespace.

This is then used by the Gateway API reconciler to synchronize relevant
ConfigMaps from BackendTLSPolicy objects into Secrets
in the configured Secrets namespace.

Includes a small refactor to move the names of indexes into
the `indexers` module (they probably should always have been there
anyway).

Signed-off-by: Nick Young <nick@isovalent.com>

youngnick committed:
gateway-api: Implement translation for BackendTLSPolicy

This commit completes the implementation of translation for
BackendTLSPolicy support.

Cilium passes the conformance tests for BackendTLSPolicy
support, although we do not currently implement the additional
BackendTLSPolicySANValidation feature. That is left for
future work, as it is a much smaller change.

It needed a few additional changes along the way:
* one of the conformance tests found a bug with the ingestion
  implementation to do with handling, so in fixing that I also wired logging down
  into the Gateway ingestion (GAMMA already had it).
* In order for our conformance report to reflect BackendTLSPolicy support
  correctly, I needed to update the Gateway API version to v1.4.1
  (which was released to correct that specific error, along with some
  other minor fixes).

Signed-off-by: Nick Young <nick@isovalent.com>

marseel commented on pkg/logging/logfields/logfields.go:
```diff
@@ -1859,4 +1859,8 @@ const (
+
+	BackendTLSPolicy = "BackendTLSPolicy"
+
+	ConfigMap = "ConfigMap"
```
nit: seems like fields for resource name are starting with lower case and have suffix "Name" at the end:
```
CiliumLocalRedirectName = "ciliumLocalRedirectPolicyName"
CiliumEgressGatewayPolicyName = "ciliumEgressGatewayPolicyName"
LRPName = "lrpName"
```

youngnick replied:
Good point, thanks! Changed.

xtineskim commented on operator/pkg/gateway-api/gateway_reconcile.go:
```diff
@@ -946,6 +962,282 @@ func (r *gatewayReconciler) setGRPCRouteStatuses(scopedLog *slog.Logger, ctx con
+// with a different generation; this means there has been a HTTPRoute update
+// between when the two lists were generated, and the whole reconciliation must be
+// restarted.
+func ContainsCommonHTTPRoute(a, b []gatewayv1.HTTPRoute) (bool, error) {
```
super minor nit, and previously discussed, should this be pulled into helpers?

youngnick replied:
Yes, done, thanks!

youngnick commented on operator/pkg/gateway-api/gateway_reconcile.go:
```diff
@@ -946,6 +962,282 @@ func (r *gatewayReconciler) setGRPCRouteStatuses(scopedLog *slog.Logger, ctx con
+			continue
+		}
+
+		scopedLog.Debug("collection", policies, collection)
```
Yes, thanks!

mhofstetter commented on operator/pkg/gateway-api/gateway.go:
```diff
@@ -324,6 +332,167 @@ func (r *gatewayReconciler) enqueueRequestForBackendService() handler.EventHandl
+		reconcileRequests := make(map[reconcile.Request]struct{})
+		btlsp, ok := o.(*gatewayv1.BackendTLSPolicy)
+		if !ok {
+			return nil
```
worth to log an error if the object isn't a `BackendTLSPolicy`? IMO we could also drop that check on the conversion and assume that this would be a developer error (if the function would be used to handle resources of another type)

mhofstetter commented on operator/pkg/gateway-api/policychecks/backendtlspolicy.go:
```diff
@@ -0,0 +1,146 @@
+	BackendTLSPolicy *gatewayv1.BackendTLSPolicy
+}
+
+func (b *BackendTLSPolicyInput) GetNamespace() string {
```
`GetNamespace` is unused

mhofstetter commented on pkg/logging/logfields/logfields.go:
```diff
@@ -1860,7 +1860,7 @@ const (
+	BackendTLSPolicyName = "backendTLSPolicyName"
 
-	ConfigMap = "ConfigMap"
+	ConfigMapName = "configMapName"
```
this (and the related changes) should be part of the 3rd commit where the logfield is introduced. this reduces the changes on the 4th commit.

youngnick commented on operator/pkg/gateway-api/gateway.go:
```diff
@@ -324,6 +332,167 @@ func (r *gatewayReconciler) enqueueRequestForBackendService() handler.EventHandl
+
+		for _, btlsp := range btlspList.Items {
+
+			serviceRefs := []string{}
```
Some of it can be reused, yeah, but, for each BackendTLSPolicy, we should be checking to see if it references the ConfigMap, and skipping that policy if not. We are not currently doing this, sigh, I missed it. I'll add and refactor. Thanks for the ping here!

youngnick commented on operator/pkg/gateway-api/gateway_reconcile.go:
```diff
@@ -946,6 +960,260 @@ func (r *gatewayReconciler) setGRPCRouteStatuses(scopedLog *slog.Logger, ctx con
+				btlsp := original.DeepCopy()
+
+				input := &policychecks.BackendTLSPolicyInput{
+					Ctx:              ctx,
```
I've stripped all of the policychecks module back to only its necessities, we can add things as needed later.

youngnick commented on operator/pkg/gateway-api/policychecks/backendtlspolicy.go:
```diff
@@ -0,0 +1,146 @@
+	"github.com/cilium/cilium/pkg/logging/logfields"
+)
+
+type BackendTLSPolicyInput struct {
```
It makes handling all of the checks and the validation functions a little easier, and copies a similar pattern we've used for the route checks. I was trying to make sure this new pattern was similar to the existing one in ways that mattered.

As I said above, I've substantially reworked and cleaned up this module, removing all the unused code. Thanks for pointing this out.

mhofstetter replied:
thanks. but  the interface `Input` is still unused. Let's remove it.

Even though the use-case is similar with the route checks - in case of the `BackerndTLSPolicy`, the "input" is only used within the method `gatewayReconciler.setBackendTLSPolicyStatuses`. No need to "force" the use of an interface type.

youngnick committed:
Update Gateway API docs for new version and BackendTLSPolicy

Signed-off-by: Nick Young <nick@isovalent.com>

mhofstetter commented on operator/pkg/gateway-api/gatewayclass_status.go:
```diff
@@ -25,7 +25,8 @@ var gatewayClassSupportedFeatures = getSupportedFeatures()
 	features.HTTPRouteParentRefPortFeature,
 	features.MeshConsumerRouteFeature,
-	features.BackendTLSPolicyFeature,
+	// features.BackendTLSPolicyFeature,
```
From the commit message i don't get why we comment out the feature `BackendTLSPolicyFeature`. Shouldn't we keep this enabled?

> Cilium passes the conformance tests for BackendTLSPolicy
support, although we do not currently implement the additional
BackendTLSPolicySANValidation feature. That is left for
future work, as it is a much smaller change.

(If this change was intentional it might be worth to add a comment)

mhofstetter replied:
great, thanks!

youngnick commented on operator/pkg/gateway-api/controller.go:
```diff
@@ -25,43 +25,6 @@ import (
 const (
 	// controllerName is the gateway controller name used in cilium.
 	controllerName = "io.cilium/gateway-controller"
-
```
I think I'll leave it for future PRs, as this is already both big and has taken a long time. Thanks for the feedback though.

youngnick commented:
/test

s-jairl commented:
Thanks for your work on this, @youngnick !
